// ngx-color/slider/esm2015/slider-swatch.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, } from '@angular/core';
let SliderSwatchComponent = class SliderSwatchComponent {
    constructor() {
        this.first = false;
        this.last = false;
        this.onClick = new EventEmitter();
    }
    ngOnChanges() {
        this.background = `hsl(${this.hsl.h}, 50%, ${this.offset * 100}%)`;
    }
    handleClick($event) {
        this.onClick.emit({
            data: {
                h: this.hsl.h,
                s: 0.5,
                l: this.offset,
                source: 'hsl',
            },
            $event,
        });
    }
};
__decorate([
    Input()
], SliderSwatchComponent.prototype, "hsl", void 0);
__decorate([
    Input()
], SliderSwatchComponent.prototype, "active", void 0);
__decorate([
    Input()
], SliderSwatchComponent.prototype, "offset", void 0);
__decorate([
    Input()
], SliderSwatchComponent.prototype, "first", void 0);
__decorate([
    Input()
], SliderSwatchComponent.prototype, "last", void 0);
__decorate([
    Output()
], SliderSwatchComponent.prototype, "onClick", void 0);
SliderSwatchComponent = __decorate([
    Component({
        selector: 'color-slider-swatch',
        template: `
  <div class="slider-swatch" [style.background]="background"
    [class.first]="first" [class.last]="last" [class.active]="active"
    (click)="handleClick($event)"
  ></div>
  `,
        changeDetection: ChangeDetectionStrategy.OnPush,
        preserveWhitespaces: false,
        styles: [`
    .slider-swatch {
      height: 12px;
      background: rgb(121, 211, 166);
      cursor: pointer;
    }
    .slider-swatch.active {
      transform: scaleY(1.8);
      border-top-right-radius: 3.6px 2px;
      border-top-left-radius: 3.6px 2px;
      border-bottom-right-radius: 3.6px 2px;
      border-bottom-left-radius: 3.6px 2px;
    }
    .slider-swatch.first {
      border-radius: 2px 0px 0px 2px;
    }
    .slider-swatch.last {
      border-radius: 0px 2px 2px 0px;
    }

  `]
    })
], SliderSwatchComponent);
export { SliderSwatchComponent };
