// ngx-color/slider/esm2015/slider.component.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { ChangeDetectionStrategy, Component, Input, NgModule, } from '@angular/core';
import { ColorWrap, HueModule, SwatchModule } from 'ngx-color';
import { SliderSwatchComponent } from './slider-swatch.component';
import { SliderSwatchesComponent } from './slider-swatches.component';
let SliderComponent = class SliderComponent extends ColorWrap {
    constructor() {
        super();
        this.pointer = {
            width: '14px',
            height: '14px',
            borderRadius: '6px',
            transform: 'translate(-7px, -2px)',
            backgroundColor: 'rgb(248, 248, 248)',
            boxShadow: '0 1px 4px 0 rgba(0, 0, 0, 0.37)',
        };
        this.radius = 2;
    }
    handlePickerChange({ data, $event }) {
        this.handleChange(data, $event);
    }
};
__decorate([
    Input()
], SliderComponent.prototype, "pointer", void 0);
__decorate([
    Input()
], SliderComponent.prototype, "radius", void 0);
SliderComponent = __decorate([
    Component({
        selector: 'color-slider',
        template: `
  <div class="slider-picker {{ className }}">
    <div class="slider-hue">
      <color-hue
        [hsl]="hsl" [radius]="radius" [pointer]="pointer"
        (onChange)="handlePickerChange($event)"
      ></color-hue>
    </div>
    <div class="slider-swatches">
      <color-slider-swatches [hsl]="hsl"
        (onClick)="handlePickerChange($event)"
      ></color-slider-swatches>
    </div>
  </div>
  `,
        changeDetection: ChangeDetectionStrategy.OnPush,
        preserveWhitespaces: false,
        styles: [`
    .slider-hue {
      height: 12px;
      position: relative;
    }
  `]
    })
], SliderComponent);
export { SliderComponent };
let ColorSliderModule = class ColorSliderModule {
};
ColorSliderModule = __decorate([
    NgModule({
        declarations: [
            SliderComponent,
            SliderSwatchComponent,
            SliderSwatchesComponent,
        ],
        exports: [SliderComponent, SliderSwatchComponent, SliderSwatchesComponent],
        imports: [CommonModule, HueModule, SwatchModule],
    })
], ColorSliderModule);
export { ColorSliderModule };
