// ngx-color/__ivy_ngcc__/alpha/fesm2015/ngx-color-alpha.js
import { __decorate } from 'tslib';
import { CommonModule } from '@angular/common';
import { Input, Component, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { ColorWrap, toState, AlphaModule, CheckboardModule } from 'ngx-color';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-color';
let AlphaPickerComponent = class AlphaPickerComponent extends ColorWrap {
    constructor() {
        super();
        /** Pixel value for picker width */
        this.width = 316;
        /** Pixel value for picker height */
        this.height = 16;
        this.direction = 'horizontal';
        this.pointer = {
            width: '18px',
            height: '18px',
            borderRadius: '50%',
            transform: 'translate(-9px, -2px)',
            boxShadow: '0 1px 4px 0 rgba(0, 0, 0, 0.37)',
        };
    }
    ngOnChanges() {
        if (this.direction === 'vertical') {
            this.pointer.transform = 'translate(-3px, -9px)';
        }
        this.setState(toState(this.color, this.oldHue));
    }
    handlePickerChange({ data, $event }) {
        this.handleChange(data, $event);
    }
};
AlphaPickerComponent.ɵfac = function AlphaPickerComponent_Factory(t) { return new (t || AlphaPickerComponent)(); };
AlphaPickerComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AlphaPickerComponent, selectors: [["color-alpha-picker"]], inputs: { width: "width", height: "height", direction: "direction" }, features: [ɵngcc0.ɵɵInheritDefinitionFeature, ɵngcc0.ɵɵNgOnChangesFeature], decls: 2, vars: 11, consts: [[3, "hsl", "rgb", "pointer", "direction", "onChange"]], template: function AlphaPickerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div");
        ɵngcc0.ɵɵelementStart(1, "color-alpha", 0);
        ɵngcc0.ɵɵlistener("onChange", function AlphaPickerComponent_Template_color_alpha_onChange_1_listener($event) { return ctx.handlePickerChange($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassMapInterpolate1("alpha-picker ", ctx.className, "");
        ɵngcc0.ɵɵstyleProp("width", ctx.width, "px")("height", ctx.height, "px");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("hsl", ctx.hsl)("rgb", ctx.rgb)("pointer", ctx.pointer)("direction", ctx.direction);
    } }, directives: [ɵngcc1.AlphaComponent], styles: [".alpha-picker[_ngcontent-%COMP%] {\n      position: relative;\n    }\n    .color-alpha[_ngcontent-%COMP%] {\n      radius: 2px;\n    }"], changeDetection: 0 });
__decorate([
    Input()
], AlphaPickerComponent.prototype, "width", void 0);
__decorate([
    Input()
], AlphaPickerComponent.prototype, "height", void 0);
__decorate([
    Input()
], AlphaPickerComponent.prototype, "direction", void 0);
let ColorAlphaModule = class ColorAlphaModule {
};
ColorAlphaModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ColorAlphaModule });
ColorAlphaModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ColorAlphaModule_Factory(t) { return new (t || ColorAlphaModule)(); }, imports: [[CommonModule, AlphaModule, CheckboardModule]] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AlphaPickerComponent, [{
        type: Component,
        args: [{
                selector: 'color-alpha-picker',
                template: `
    <div class="alpha-picker {{ className }}"
      [style.width.px]="width" [style.height.px]="height">
      <color-alpha
        [hsl]="hsl"
        [rgb]="rgb"
        [pointer]="pointer"
        [direction]="direction"
        (onChange)="handlePickerChange($event)"
      ></color-alpha>
    </div>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush,
                preserveWhitespaces: false,
                styles: [`
    .alpha-picker {
      position: relative;
    }
    .color-alpha {
      radius: 2px;
    }
  `]
            }]
    }], function () { return []; }, { width: [{
            type: Input
        }], height: [{
            type: Input
        }], direction: [{
            type: Input
        }] }); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ColorAlphaModule, { declarations: function () { return [AlphaPickerComponent]; }, imports: function () { return [CommonModule, AlphaModule, CheckboardModule]; }, exports: function () { return [AlphaPickerComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ColorAlphaModule, [{
        type: NgModule,
        args: [{
                declarations: [AlphaPickerComponent],
                exports: [AlphaPickerComponent],
                imports: [CommonModule, AlphaModule, CheckboardModule]
            }]
    }], null, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { AlphaPickerComponent, ColorAlphaModule };
