// @ng-matero/extensions/esm5/split-pane/interface.js
/**
 * @fileoverview added by tsickle
 * Generated from: interface.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function MtxSplitPoint() { }
if (false) {
    /** @type {?} */
    MtxSplitPoint.prototype.x;
    /** @type {?} */
    MtxSplitPoint.prototype.y;
}
/**
 * @record
 */
export function MtxSplitArea() { }
if (false) {
    /** @type {?} */
    MtxSplitArea.prototype.component;
    /** @type {?} */
    MtxSplitArea.prototype.order;
    /** @type {?} */
    MtxSplitArea.prototype.size;
    /** @type {?} */
    MtxSplitArea.prototype.minSize;
    /** @type {?} */
    MtxSplitArea.prototype.maxSize;
}
/**
 * @record
 */
export function MtxSplitSnapshot() { }
if (false) {
    /** @type {?} */
    MtxSplitSnapshot.prototype.gutterNum;
    /** @type {?} */
    MtxSplitSnapshot.prototype.allAreasSizePixel;
    /** @type {?} */
    MtxSplitSnapshot.prototype.allInvolvedAreasSizePercent;
    /** @type {?} */
    MtxSplitSnapshot.prototype.lastSteppedOffset;
    /** @type {?} */
    MtxSplitSnapshot.prototype.areasBeforeGutter;
    /** @type {?} */
    MtxSplitSnapshot.prototype.areasAfterGutter;
}
/**
 * @record
 */
export function MtxSplitAreaSnapshot() { }
if (false) {
    /** @type {?} */
    MtxSplitAreaSnapshot.prototype.area;
    /** @type {?} */
    MtxSplitAreaSnapshot.prototype.sizePixelAtStart;
    /** @type {?} */
    MtxSplitAreaSnapshot.prototype.sizePercentAtStart;
}
/**
 * @record
 */
export function MtxSplitSideAbsorptionCapacity() { }
if (false) {
    /** @type {?} */
    MtxSplitSideAbsorptionCapacity.prototype.remain;
    /** @type {?} */
    MtxSplitSideAbsorptionCapacity.prototype.list;
}
/**
 * @record
 */
export function MtxSplitAreaAbsorptionCapacity() { }
if (false) {
    /** @type {?} */
    MtxSplitAreaAbsorptionCapacity.prototype.areaSnapshot;
    /** @type {?} */
    MtxSplitAreaAbsorptionCapacity.prototype.pixelAbsorb;
    /** @type {?} */
    MtxSplitAreaAbsorptionCapacity.prototype.percentAfterAbsorption;
    /** @type {?} */
    MtxSplitAreaAbsorptionCapacity.prototype.pixelRemain;
}
/**
 * @record
 */
export function MtxSplitOutputData() { }
if (false) {
    /** @type {?} */
    MtxSplitOutputData.prototype.gutterNum;
    /** @type {?} */
    MtxSplitOutputData.prototype.sizes;
}
/**
 * @record
 */
export function MtxSplitOutputAreaSizes() { }
