// ngx-color/hue/fesm5/ngx-color-hue.js
import { __extends, __decorate } from 'tslib';
import { CommonModule } from '@angular/common';
import { Input, Component, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { toState, ColorWrap, HueModule } from 'ngx-color';

var HuePickerComponent = /** @class */ (function (_super) {
    __extends(HuePickerComponent, _super);
    function HuePickerComponent() {
        var _this = _super.call(this) || this;
        /** Pixel value for picker width */
        _this.width = 316;
        /** Pixel value for picker height */
        _this.height = 16;
        _this.radius = 2;
        _this.direction = 'horizontal';
        _this.pointer = {
            width: '18px',
            height: '18px',
            borderRadius: '50%',
            transform: 'translate(-9px, -2px)',
            backgroundColor: 'rgb(248, 248, 248)',
            boxShadow: '0 1px 4px 0 rgba(0, 0, 0, 0.37)',
        };
        return _this;
    }
    HuePickerComponent.prototype.ngOnChanges = function () {
        if (this.direction === 'vertical') {
            this.pointer.transform = 'translate(-3px, -9px)';
        }
        this.setState(toState(this.color, this.oldHue));
    };
    HuePickerComponent.prototype.handlePickerChange = function (_a) {
        var data = _a.data, $event = _a.$event;
        this.handleChange({ a: 1, h: data.h, l: 0.5, s: 1 }, $event);
    };
    __decorate([
        Input()
    ], HuePickerComponent.prototype, "width", void 0);
    __decorate([
        Input()
    ], HuePickerComponent.prototype, "height", void 0);
    __decorate([
        Input()
    ], HuePickerComponent.prototype, "radius", void 0);
    __decorate([
        Input()
    ], HuePickerComponent.prototype, "direction", void 0);
    HuePickerComponent = __decorate([
        Component({
            selector: 'color-hue-picker',
            template: "\n  <div class=\"hue-picker {{ className }}\"\n    [style.width.px]=\"width\" [style.height.px]=\"height\"\n  >\n    <color-hue [hsl]=\"hsl\" [pointer]=\"pointer\"\n      [direction]=\"direction\" [radius]=\"radius\"\n      (onChange)=\"handlePickerChange($event)\"\n    ></color-hue>\n  </div>\n  ",
            changeDetection: ChangeDetectionStrategy.OnPush,
            preserveWhitespaces: false,
            styles: ["\n    .hue-picker {\n      position: relative;\n    }\n  "]
        })
    ], HuePickerComponent);
    return HuePickerComponent;
}(ColorWrap));
var ColorHueModule = /** @class */ (function () {
    function ColorHueModule() {
    }
    ColorHueModule = __decorate([
        NgModule({
            declarations: [HuePickerComponent],
            exports: [HuePickerComponent],
            imports: [CommonModule, HueModule],
        })
    ], ColorHueModule);
    return ColorHueModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { ColorHueModule, HuePickerComponent };
