// @ng-matero/extensions/esm5/select/option.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: option.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ElementRef, Input, } from '@angular/core';
import { Subject } from 'rxjs';
var MtxOptionComponent = /** @class */ (function () {
    function MtxOptionComponent(elementRef) {
        this.elementRef = elementRef;
        this.stateChange$ = new Subject();
        this._disabled = false;
    }
    Object.defineProperty(MtxOptionComponent.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = this._isDisabled(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MtxOptionComponent.prototype, "label", {
        get: /**
         * @return {?}
         */
        function () {
            return (this.elementRef.nativeElement.textContent || '').trim();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} changes
     * @return {?}
     */
    MtxOptionComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.disabled) {
            this.stateChange$.next({
                value: this.value,
                disabled: this._disabled,
            });
        }
    };
    /**
     * @return {?}
     */
    MtxOptionComponent.prototype.ngAfterViewChecked = /**
     * @return {?}
     */
    function () {
        if (this.label !== this._previousLabel) {
            this._previousLabel = this.label;
            this.stateChange$.next({
                value: this.value,
                disabled: this._disabled,
                label: this.elementRef.nativeElement.innerHTML,
            });
        }
    };
    /**
     * @return {?}
     */
    MtxOptionComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.stateChange$.complete();
    };
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    MtxOptionComponent.prototype._isDisabled = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value != null && "" + value !== 'false';
    };
    MtxOptionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mtx-option',
                    exportAs: 'mtxOption',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: "<ng-content></ng-content>"
                }] }
    ];
    /** @nocollapse */
    MtxOptionComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    MtxOptionComponent.propDecorators = {
        value: [{ type: Input }],
        disabled: [{ type: Input }]
    };
    return MtxOptionComponent;
}());
export { MtxOptionComponent };
if (false) {
    /** @type {?} */
    MtxOptionComponent.prototype.value;
    /** @type {?} */
    MtxOptionComponent.prototype.stateChange$;
    /**
     * @type {?}
     * @private
     */
    MtxOptionComponent.prototype._disabled;
    /**
     * @type {?}
     * @private
     */
    MtxOptionComponent.prototype._previousLabel;
    /** @type {?} */
    MtxOptionComponent.prototype.elementRef;
}
