// @ng-matero/extensions/esm5/select/templates.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: templates.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, TemplateRef } from '@angular/core';
var MtxSelectOptionTemplateDirective = /** @class */ (function () {
    function MtxSelectOptionTemplateDirective(template) {
        this.template = template;
    }
    MtxSelectOptionTemplateDirective.decorators = [
        { type: Directive, args: [{ selector: '[ng-option-tmp]' },] }
    ];
    /** @nocollapse */
    MtxSelectOptionTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return MtxSelectOptionTemplateDirective;
}());
export { MtxSelectOptionTemplateDirective };
if (false) {
    /** @type {?} */
    MtxSelectOptionTemplateDirective.prototype.template;
}
var MtxSelectOptgroupTemplateDirective = /** @class */ (function () {
    function MtxSelectOptgroupTemplateDirective(template) {
        this.template = template;
    }
    MtxSelectOptgroupTemplateDirective.decorators = [
        { type: Directive, args: [{ selector: '[ng-optgroup-tmp]' },] }
    ];
    /** @nocollapse */
    MtxSelectOptgroupTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return MtxSelectOptgroupTemplateDirective;
}());
export { MtxSelectOptgroupTemplateDirective };
if (false) {
    /** @type {?} */
    MtxSelectOptgroupTemplateDirective.prototype.template;
}
var MtxSelectLabelTemplateDirective = /** @class */ (function () {
    function MtxSelectLabelTemplateDirective(template) {
        this.template = template;
    }
    MtxSelectLabelTemplateDirective.decorators = [
        { type: Directive, args: [{ selector: '[ng-label-tmp]' },] }
    ];
    /** @nocollapse */
    MtxSelectLabelTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return MtxSelectLabelTemplateDirective;
}());
export { MtxSelectLabelTemplateDirective };
if (false) {
    /** @type {?} */
    MtxSelectLabelTemplateDirective.prototype.template;
}
var MtxSelectMultiLabelTemplateDirective = /** @class */ (function () {
    function MtxSelectMultiLabelTemplateDirective(template) {
        this.template = template;
    }
    MtxSelectMultiLabelTemplateDirective.decorators = [
        { type: Directive, args: [{ selector: '[ng-multi-label-tmp]' },] }
    ];
    /** @nocollapse */
    MtxSelectMultiLabelTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return MtxSelectMultiLabelTemplateDirective;
}());
export { MtxSelectMultiLabelTemplateDirective };
if (false) {
    /** @type {?} */
    MtxSelectMultiLabelTemplateDirective.prototype.template;
}
var MtxSelectHeaderTemplateDirective = /** @class */ (function () {
    function MtxSelectHeaderTemplateDirective(template) {
        this.template = template;
    }
    MtxSelectHeaderTemplateDirective.decorators = [
        { type: Directive, args: [{ selector: '[ng-header-tmp]' },] }
    ];
    /** @nocollapse */
    MtxSelectHeaderTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return MtxSelectHeaderTemplateDirective;
}());
export { MtxSelectHeaderTemplateDirective };
if (false) {
    /** @type {?} */
    MtxSelectHeaderTemplateDirective.prototype.template;
}
var MtxSelectFooterTemplateDirective = /** @class */ (function () {
    function MtxSelectFooterTemplateDirective(template) {
        this.template = template;
    }
    MtxSelectFooterTemplateDirective.decorators = [
        { type: Directive, args: [{ selector: '[ng-footer-tmp]' },] }
    ];
    /** @nocollapse */
    MtxSelectFooterTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return MtxSelectFooterTemplateDirective;
}());
export { MtxSelectFooterTemplateDirective };
if (false) {
    /** @type {?} */
    MtxSelectFooterTemplateDirective.prototype.template;
}
var MtxSelectNotFoundTemplateDirective = /** @class */ (function () {
    function MtxSelectNotFoundTemplateDirective(template) {
        this.template = template;
    }
    MtxSelectNotFoundTemplateDirective.decorators = [
        { type: Directive, args: [{ selector: '[ng-notfound-tmp]' },] }
    ];
    /** @nocollapse */
    MtxSelectNotFoundTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return MtxSelectNotFoundTemplateDirective;
}());
export { MtxSelectNotFoundTemplateDirective };
if (false) {
    /** @type {?} */
    MtxSelectNotFoundTemplateDirective.prototype.template;
}
var MtxSelectTypeToSearchTemplateDirective = /** @class */ (function () {
    function MtxSelectTypeToSearchTemplateDirective(template) {
        this.template = template;
    }
    MtxSelectTypeToSearchTemplateDirective.decorators = [
        { type: Directive, args: [{ selector: '[ng-typetosearch-tmp]' },] }
    ];
    /** @nocollapse */
    MtxSelectTypeToSearchTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return MtxSelectTypeToSearchTemplateDirective;
}());
export { MtxSelectTypeToSearchTemplateDirective };
if (false) {
    /** @type {?} */
    MtxSelectTypeToSearchTemplateDirective.prototype.template;
}
var MtxSelectLoadingTextTemplateDirective = /** @class */ (function () {
    function MtxSelectLoadingTextTemplateDirective(template) {
        this.template = template;
    }
    MtxSelectLoadingTextTemplateDirective.decorators = [
        { type: Directive, args: [{ selector: '[ng-loadingtext-tmp]' },] }
    ];
    /** @nocollapse */
    MtxSelectLoadingTextTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return MtxSelectLoadingTextTemplateDirective;
}());
export { MtxSelectLoadingTextTemplateDirective };
if (false) {
    /** @type {?} */
    MtxSelectLoadingTextTemplateDirective.prototype.template;
}
var MtxSelectTagTemplateDirective = /** @class */ (function () {
    function MtxSelectTagTemplateDirective(template) {
        this.template = template;
    }
    MtxSelectTagTemplateDirective.decorators = [
        { type: Directive, args: [{ selector: '[ng-tag-tmp]' },] }
    ];
    /** @nocollapse */
    MtxSelectTagTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return MtxSelectTagTemplateDirective;
}());
export { MtxSelectTagTemplateDirective };
if (false) {
    /** @type {?} */
    MtxSelectTagTemplateDirective.prototype.template;
}
var MtxSelectLoadingSpinnerTemplateDirective = /** @class */ (function () {
    function MtxSelectLoadingSpinnerTemplateDirective(template) {
        this.template = template;
    }
    MtxSelectLoadingSpinnerTemplateDirective.decorators = [
        { type: Directive, args: [{ selector: '[ng-loadingspinner-tmp]' },] }
    ];
    /** @nocollapse */
    MtxSelectLoadingSpinnerTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return MtxSelectLoadingSpinnerTemplateDirective;
}());
export { MtxSelectLoadingSpinnerTemplateDirective };
if (false) {
    /** @type {?} */
    MtxSelectLoadingSpinnerTemplateDirective.prototype.template;
}
