// ngx-color/block/esm2015/block-swatches.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
let BlockSwatchesComponent = class BlockSwatchesComponent {
    constructor() {
        this.onClick = new EventEmitter();
        this.onSwatchHover = new EventEmitter();
    }
    ngOnInit() {
        this.swatchStyle = {
            width: '22px',
            height: '22px',
            float: 'left',
            marginRight: '10px',
            marginBottom: '10px',
            borderRadius: '4px',
        };
    }
    handleClick({ hex, $event }) {
        this.onClick.emit({ hex, $event });
    }
    focusStyle(c) {
        return {
            boxShadow: `${c} 0 0 4px`,
        };
    }
};
__decorate([
    Input()
], BlockSwatchesComponent.prototype, "colors", void 0);
__decorate([
    Output()
], BlockSwatchesComponent.prototype, "onClick", void 0);
__decorate([
    Output()
], BlockSwatchesComponent.prototype, "onSwatchHover", void 0);
BlockSwatchesComponent = __decorate([
    Component({
        selector: 'color-block-swatches',
        template: `
    <div class="block-swatches">
      <color-swatch
        *ngFor="let c of colors"
        [color]="c"
        [style]="swatchStyle"
        [focusStyle]="focusStyle(c)"
        (onClick)="handleClick($event)"
        (onHover)="onSwatchHover.emit($event)"
      ></color-swatch>
      <div class="clear"></div>
    </div>
  `,
        styles: [`
    .block-swatches {
      margin-right: -10px;
    }
    .clear {
      clear: both;
    }
  `]
    })
], BlockSwatchesComponent);
export { BlockSwatchesComponent };
