// ngx-color/swatches/esm5/swatches-group.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, } from '@angular/core';
var SwatchesGroupComponent = /** @class */ (function () {
    function SwatchesGroupComponent() {
        this.onClick = new EventEmitter();
        this.onSwatchHover = new EventEmitter();
    }
    __decorate([
        Input()
    ], SwatchesGroupComponent.prototype, "group", void 0);
    __decorate([
        Input()
    ], SwatchesGroupComponent.prototype, "active", void 0);
    __decorate([
        Output()
    ], SwatchesGroupComponent.prototype, "onClick", void 0);
    __decorate([
        Output()
    ], SwatchesGroupComponent.prototype, "onSwatchHover", void 0);
    SwatchesGroupComponent = __decorate([
        Component({
            selector: 'color-swatches-group',
            template: "\n  <div class=\"swatches-group\">\n    <color-swatches-color\n      *ngFor=\"let color of group; let idx = index;\"\n      [active]=\"color.toLowerCase() === active\"\n      [color]=\"color\"\n      [first]=\"idx === 0\"\n      [last]=\"idx === group.length - 1\"\n      (onClick)=\"onClick.emit($event)\"\n    >\n    </color-swatches-color>\n  </div>\n  ",
            changeDetection: ChangeDetectionStrategy.OnPush,
            preserveWhitespaces: false,
            styles: ["\n    .swatches-group {\n      padding-bottom: 10px;\n      width: 40px;\n      float: left;\n      margin-right: 10px;\n    }\n  "]
        })
    ], SwatchesGroupComponent);
    return SwatchesGroupComponent;
}());
export { SwatchesGroupComponent };
