// ngx-color/alpha/esm2015/alpha-picker.component.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { ChangeDetectionStrategy, Component, Input, NgModule, } from '@angular/core';
import { AlphaModule, CheckboardModule, ColorWrap, toState } from 'ngx-color';
let AlphaPickerComponent = class AlphaPickerComponent extends ColorWrap {
    constructor() {
        super();
        /** Pixel value for picker width */
        this.width = 316;
        /** Pixel value for picker height */
        this.height = 16;
        this.direction = 'horizontal';
        this.pointer = {
            width: '18px',
            height: '18px',
            borderRadius: '50%',
            transform: 'translate(-9px, -2px)',
            boxShadow: '0 1px 4px 0 rgba(0, 0, 0, 0.37)',
        };
    }
    ngOnChanges() {
        if (this.direction === 'vertical') {
            this.pointer.transform = 'translate(-3px, -9px)';
        }
        this.setState(toState(this.color, this.oldHue));
    }
    handlePickerChange({ data, $event }) {
        this.handleChange(data, $event);
    }
};
__decorate([
    Input()
], AlphaPickerComponent.prototype, "width", void 0);
__decorate([
    Input()
], AlphaPickerComponent.prototype, "height", void 0);
__decorate([
    Input()
], AlphaPickerComponent.prototype, "direction", void 0);
AlphaPickerComponent = __decorate([
    Component({
        selector: 'color-alpha-picker',
        template: `
    <div class="alpha-picker {{ className }}"
      [style.width.px]="width" [style.height.px]="height">
      <color-alpha
        [hsl]="hsl"
        [rgb]="rgb"
        [pointer]="pointer"
        [direction]="direction"
        (onChange)="handlePickerChange($event)"
      ></color-alpha>
    </div>
  `,
        changeDetection: ChangeDetectionStrategy.OnPush,
        preserveWhitespaces: false,
        styles: [`
    .alpha-picker {
      position: relative;
    }
    .color-alpha {
      radius: 2px;
    }
  `]
    })
], AlphaPickerComponent);
export { AlphaPickerComponent };
let ColorAlphaModule = class ColorAlphaModule {
};
ColorAlphaModule = __decorate([
    NgModule({
        declarations: [AlphaPickerComponent],
        exports: [AlphaPickerComponent],
        imports: [CommonModule, AlphaModule, CheckboardModule],
    })
], ColorAlphaModule);
export { ColorAlphaModule };
