// ngx-color/shade/fesm2015/ngx-color-shade.js
import { __decorate } from 'tslib';
import { CommonModule } from '@angular/common';
import { Input, Component, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { ColorWrap, toState, ShadeModule } from 'ngx-color';

let ShadeSliderComponent = class ShadeSliderComponent extends ColorWrap {
    constructor() {
        super();
        /** Pixel value for picker width */
        this.width = 316;
        /** Pixel value for picker height */
        this.height = 16;
        this.pointer = {
            width: '18px',
            height: '18px',
            borderRadius: '50%',
            transform: 'translate(-9px, -2px)',
            boxShadow: '0 1px 4px 0 rgba(0, 0, 0, 0.37)',
        };
    }
    ngOnChanges() {
        this.setState(toState(this.color, this.oldHue));
    }
    handlePickerChange({ data, $event }) {
        this.handleChange(data, $event);
    }
};
__decorate([
    Input()
], ShadeSliderComponent.prototype, "width", void 0);
__decorate([
    Input()
], ShadeSliderComponent.prototype, "height", void 0);
ShadeSliderComponent = __decorate([
    Component({
        selector: 'color-shade-picker',
        template: `
    <div class="shade-slider {{ className || '' }}"
      [style.width.px]="width" [style.height.px]="height">
      <color-shade
        [hsl]="hsl"
        [rgb]="rgb"
        [pointer]="pointer"
        (onChange)="handlePickerChange($event)"
      ></color-shade>
    </div>
  `,
        changeDetection: ChangeDetectionStrategy.OnPush,
        preserveWhitespaces: false,
        styles: [`
    .shade-slider {
      position: relative;
    }
  `]
    })
], ShadeSliderComponent);
let ColorShadeModule = class ColorShadeModule {
};
ColorShadeModule = __decorate([
    NgModule({
        declarations: [ShadeSliderComponent],
        exports: [ShadeSliderComponent],
        imports: [CommonModule, ShadeModule],
    })
], ColorShadeModule);

/**
 * Generated bundle index. Do not edit.
 */

export { ColorShadeModule, ShadeSliderComponent };
