// @ng-matero/extensions/esm2015/dialog/dialog.js
/**
 * @fileoverview added by tsickle
 * Generated from: dialog.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { MtxDialogComponent } from './dialog.component';
const ɵ0 = /**
 * @return {?}
 */
() => { }, ɵ1 = /**
 * @return {?}
 */
() => { };
/** @type {?} */
const defaults = {
    title: '',
    description: '',
    buttons: [
        {
            type: '',
            text: 'CLOSE',
            onClick: (ɵ0),
        },
        {
            type: 'warn',
            text: 'OK',
            onClick: (ɵ1),
        },
    ],
    width: '300px',
};
export class MtxDialog {
    /**
     * @param {?} dialog
     */
    constructor(dialog) {
        this.dialog = dialog;
    }
    /**
     * @param {?=} componentOrTemplateRef
     * @param {?=} config
     * @return {?}
     */
    originalOpen(componentOrTemplateRef = MtxDialogComponent, config) {
        return this.dialog.open(componentOrTemplateRef, config);
    }
    /**
     * @param {?} config
     * @param {?=} componentOrTemplateRef
     * @return {?}
     */
    open(config, componentOrTemplateRef = MtxDialogComponent) {
        /** @type {?} */
        const data = Object.assign({}, defaults, config);
        return this.dialog.open(componentOrTemplateRef, Object.assign(Object.assign({}, data), { data }));
    }
    /**
     * @param {?} title
     * @param {?=} onOk
     * @return {?}
     */
    alert(title, onOk = (/**
     * @return {?}
     */
    () => { })) {
        this.open({
            title,
            buttons: [
                {
                    type: 'warn',
                    text: 'OK',
                    onClick: (/**
                     * @return {?}
                     */
                    () => onOk()),
                },
            ],
            disableClose: true,
        });
    }
    /**
     * @param {?} title
     * @param {?=} onOk
     * @param {?=} onClose
     * @return {?}
     */
    confirm(title, onOk = (/**
     * @return {?}
     */
    () => { }), onClose = (/**
     * @return {?}
     */
    () => { })) {
        this.open({
            title,
            buttons: [
                {
                    type: '',
                    text: 'CLOSE',
                    onClick: (/**
                     * @return {?}
                     */
                    () => onClose()),
                },
                {
                    type: 'warn',
                    text: 'OK',
                    onClick: (/**
                     * @return {?}
                     */
                    () => onOk()),
                },
            ],
            disableClose: true,
        });
    }
}
MtxDialog.decorators = [
    { type: Injectable }
];
/** @nocollapse */
MtxDialog.ctorParameters = () => [
    { type: MatDialog }
];
if (false) {
    /** @type {?} */
    MtxDialog.prototype.dialog;
}
export { ɵ0, ɵ1 };
