// @ng-matero/extensions/esm5/data-grid/cell-selection.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: cell-selection.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, EventEmitter, HostBinding, HostListener, Input, Output, } from '@angular/core';
import { MtxGridComponent } from './grid.component';
var MtxGridCellSelectionDirective = /** @class */ (function () {
    function MtxGridCellSelectionDirective(_dataGrid) {
        this._dataGrid = _dataGrid;
        this._selected = false;
        this.shiftKeyPressed = false;
        this.ctrlKeyPressed = false;
        this.cellSelectionChange = new EventEmitter();
    }
    Object.defineProperty(MtxGridCellSelectionDirective.prototype, "selected", {
        get: /**
         * @return {?}
         */
        function () {
            return this._selected;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MtxGridCellSelectionDirective.prototype, "matSelectableRowData", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value !== this._rowData) {
                this._rowData = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    MtxGridCellSelectionDirective.prototype.onClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.ctrlKeyPressed = event.ctrlKey;
        this.shiftKeyPressed = event.shiftKey;
        if (this._dataGrid.cellSelectable) {
            this.select();
        }
    };
    /**
     * @return {?}
     */
    MtxGridCellSelectionDirective.prototype.select = /**
     * @return {?}
     */
    function () {
        this._selected = true;
        this.cellSelectionChange.emit(this);
    };
    /**
     * @return {?}
     */
    MtxGridCellSelectionDirective.prototype.deselect = /**
     * @return {?}
     */
    function () {
        this._selected = false;
        this.cellSelectionChange.emit(this);
    };
    /**
     * @return {?}
     */
    MtxGridCellSelectionDirective.prototype.toggle = /**
     * @return {?}
     */
    function () {
        this._selected = !this._selected;
        this.cellSelectionChange.emit(this);
    };
    MtxGridCellSelectionDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[mtx-selectable-cell]',
                },] }
    ];
    /** @nocollapse */
    MtxGridCellSelectionDirective.ctorParameters = function () { return [
        { type: MtxGridComponent }
    ]; };
    MtxGridCellSelectionDirective.propDecorators = {
        selected: [{ type: HostBinding, args: ['class.selected',] }],
        matSelectableRowData: [{ type: Input }],
        cellSelectionChange: [{ type: Output }],
        onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return MtxGridCellSelectionDirective;
}());
export { MtxGridCellSelectionDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    MtxGridCellSelectionDirective.prototype._selected;
    /**
     * @type {?}
     * @private
     */
    MtxGridCellSelectionDirective.prototype._rowData;
    /** @type {?} */
    MtxGridCellSelectionDirective.prototype.shiftKeyPressed;
    /** @type {?} */
    MtxGridCellSelectionDirective.prototype.ctrlKeyPressed;
    /** @type {?} */
    MtxGridCellSelectionDirective.prototype.cellSelectionChange;
    /**
     * @type {?}
     * @private
     */
    MtxGridCellSelectionDirective.prototype._dataGrid;
}
