// @ng-matero/extensions/esm5/data-grid/expansion-toggle.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: expansion-toggle.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, EventEmitter, HostBinding, HostListener, Input, Output, TemplateRef, } from '@angular/core';
var MtxGridExpansionToggleDirective = /** @class */ (function () {
    function MtxGridExpansionToggleDirective() {
        this._opened = false;
        this.toggleChange = new EventEmitter();
    }
    Object.defineProperty(MtxGridExpansionToggleDirective.prototype, "expended", {
        get: /**
         * @return {?}
         */
        function () {
            return this._opened;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MtxGridExpansionToggleDirective.prototype, "expandableRow", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value !== this._row) {
                this._row = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MtxGridExpansionToggleDirective.prototype, "template", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value !== this._tplRef) {
                this._tplRef = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    MtxGridExpansionToggleDirective.prototype.onClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        event.stopPropagation();
        this.toggle();
    };
    /**
     * @return {?}
     */
    MtxGridExpansionToggleDirective.prototype.toggle = /**
     * @return {?}
     */
    function () {
        this._opened = !this._opened;
        this.toggleChange.emit(this);
    };
    MtxGridExpansionToggleDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[mtx-expansion-toggle]',
                },] }
    ];
    /** @nocollapse */
    MtxGridExpansionToggleDirective.ctorParameters = function () { return []; };
    MtxGridExpansionToggleDirective.propDecorators = {
        expended: [{ type: HostBinding, args: ['class.expanded',] }],
        expandableRow: [{ type: Input }],
        template: [{ type: Input, args: ['expansionRowTpl',] }],
        toggleChange: [{ type: Output }],
        onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return MtxGridExpansionToggleDirective;
}());
export { MtxGridExpansionToggleDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    MtxGridExpansionToggleDirective.prototype._opened;
    /**
     * @type {?}
     * @private
     */
    MtxGridExpansionToggleDirective.prototype._row;
    /**
     * @type {?}
     * @private
     */
    MtxGridExpansionToggleDirective.prototype._tplRef;
    /** @type {?} */
    MtxGridExpansionToggleDirective.prototype.toggleChange;
}
