// ngx-color/esm5/checkboard.component.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { ChangeDetectionStrategy, Component, Input, NgModule, } from '@angular/core';
import { getCheckerboard } from './helpers/checkboard';
var CheckboardComponent = /** @class */ (function () {
    function CheckboardComponent() {
        this.white = 'transparent';
        this.size = 8;
        this.grey = 'rgba(0,0,0,.08)';
    }
    CheckboardComponent.prototype.ngOnInit = function () {
        var background = getCheckerboard(this.white, this.grey, this.size);
        this.gridStyles = {
            borderRadius: this.borderRadius,
            boxShadow: this.boxShadow,
            background: "url(" + background + ") center left",
        };
    };
    __decorate([
        Input()
    ], CheckboardComponent.prototype, "white", void 0);
    __decorate([
        Input()
    ], CheckboardComponent.prototype, "size", void 0);
    __decorate([
        Input()
    ], CheckboardComponent.prototype, "grey", void 0);
    __decorate([
        Input()
    ], CheckboardComponent.prototype, "boxShadow", void 0);
    __decorate([
        Input()
    ], CheckboardComponent.prototype, "borderRadius", void 0);
    CheckboardComponent = __decorate([
        Component({
            selector: 'color-checkboard',
            template: "<div class=\"grid\" [ngStyle]=\"gridStyles\"></div>",
            preserveWhitespaces: false,
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: ["\n  .grid {\n    top: 0px;\n    right: 0px;\n    bottom: 0px;\n    left: 0px;\n    position: absolute;\n  }\n  "]
        })
    ], CheckboardComponent);
    return CheckboardComponent;
}());
export { CheckboardComponent };
var CheckboardModule = /** @class */ (function () {
    function CheckboardModule() {
    }
    CheckboardModule = __decorate([
        NgModule({
            declarations: [CheckboardComponent],
            exports: [CheckboardComponent],
            imports: [CommonModule],
        })
    ], CheckboardModule);
    return CheckboardModule;
}());
export { CheckboardModule };
