// ngx-color/esm5/swatch.component.js
import { __assign, __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { ChangeDetectionStrategy, Component, EventEmitter, Input, NgModule, Output, } from '@angular/core';
import { CheckboardModule } from './checkboard.component';
var SwatchComponent = /** @class */ (function () {
    function SwatchComponent() {
        this.style = {};
        this.focusStyle = {};
        this.onClick = new EventEmitter();
        this.onHover = new EventEmitter();
        this.divStyles = {};
        this.focusStyles = {};
        this.inFocus = false;
    }
    SwatchComponent.prototype.ngOnInit = function () {
        this.divStyles = __assign({ background: this.color, height: '100%', width: '100%', cursor: 'pointer', position: 'relative', outline: 'none' }, this.style);
    };
    SwatchComponent.prototype.currentStyles = function () {
        this.focusStyles = __assign(__assign({}, this.divStyles), this.focusStyle);
        return this.focus || this.inFocus ? this.focusStyles : this.divStyles;
    };
    SwatchComponent.prototype.handleFocusOut = function () {
        this.inFocus = false;
    };
    SwatchComponent.prototype.handleFocus = function () {
        this.inFocus = true;
    };
    SwatchComponent.prototype.handleHover = function (hex, $event) {
        this.onHover.emit({ hex: hex, $event: $event });
    };
    SwatchComponent.prototype.handleClick = function (hex, $event) {
        this.onClick.emit({ hex: hex, $event: $event });
    };
    __decorate([
        Input()
    ], SwatchComponent.prototype, "color", void 0);
    __decorate([
        Input()
    ], SwatchComponent.prototype, "style", void 0);
    __decorate([
        Input()
    ], SwatchComponent.prototype, "focusStyle", void 0);
    __decorate([
        Input()
    ], SwatchComponent.prototype, "focus", void 0);
    __decorate([
        Output()
    ], SwatchComponent.prototype, "onClick", void 0);
    __decorate([
        Output()
    ], SwatchComponent.prototype, "onHover", void 0);
    SwatchComponent = __decorate([
        Component({
            selector: 'color-swatch',
            template: "\n    <div\n      class=\"swatch\"\n      [ngStyle]=\"currentStyles()\"\n      [attr.title]=\"color\"\n      (click)=\"handleClick(color, $event)\"\n      (keydown.enter)=\"handleClick(color, $event)\"\n      (focus)=\"handleFocus()\"\n      (blur)=\"handleFocusOut()\"\n      (mouseover)=\"handleHover(color, $event)\"\n      tabindex=\"0\"\n    >\n      <ng-content></ng-content>\n      <color-checkboard\n        *ngIf=\"color === 'transparent'\"\n        boxShadow=\"inset 0 0 0 1px rgba(0,0,0,0.1)\"\n      ></color-checkboard>\n    </div>\n  ",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], SwatchComponent);
    return SwatchComponent;
}());
export { SwatchComponent };
var SwatchModule = /** @class */ (function () {
    function SwatchModule() {
    }
    SwatchModule = __decorate([
        NgModule({
            declarations: [SwatchComponent],
            exports: [SwatchComponent],
            imports: [CommonModule, CheckboardModule],
        })
    ], SwatchModule);
    return SwatchModule;
}());
export { SwatchModule };
