// @ng-matero/extensions/esm5/dialog/dialog.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: dialog.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatDialogModule } from '@angular/material/dialog';
import { MatButtonModule } from '@angular/material/button';
import { MtxDialogComponent } from './dialog.component';
import { MtxDialog } from './dialog';
var MtxDialogModule = /** @class */ (function () {
    function MtxDialogModule() {
    }
    MtxDialogModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, MatButtonModule, MatDialogModule],
                    exports: [MtxDialogComponent],
                    declarations: [MtxDialogComponent],
                    providers: [MtxDialog],
                    entryComponents: [MtxDialogComponent],
                },] }
    ];
    return MtxDialogModule;
}());
export { MtxDialogModule };
