// @ng-matero/extensions/esm5/dialog/dialog.js
/**
 * @fileoverview added by tsickle
 * Generated from: dialog.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { __assign } from "tslib";
import { Injectable } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { MtxDialogComponent } from './dialog.component';
var ɵ0 = /**
 * @return {?}
 */
function () { }, ɵ1 = /**
 * @return {?}
 */
function () { };
/** @type {?} */
var defaults = {
    title: '',
    description: '',
    buttons: [
        {
            type: '',
            text: 'CLOSE',
            onClick: (ɵ0),
        },
        {
            type: 'warn',
            text: 'OK',
            onClick: (ɵ1),
        },
    ],
    width: '300px',
};
var MtxDialog = /** @class */ (function () {
    function MtxDialog(dialog) {
        this.dialog = dialog;
    }
    /**
     * @param {?=} componentOrTemplateRef
     * @param {?=} config
     * @return {?}
     */
    MtxDialog.prototype.originalOpen = /**
     * @param {?=} componentOrTemplateRef
     * @param {?=} config
     * @return {?}
     */
    function (componentOrTemplateRef, config) {
        if (componentOrTemplateRef === void 0) { componentOrTemplateRef = MtxDialogComponent; }
        return this.dialog.open(componentOrTemplateRef, config);
    };
    /**
     * @param {?} config
     * @param {?=} componentOrTemplateRef
     * @return {?}
     */
    MtxDialog.prototype.open = /**
     * @param {?} config
     * @param {?=} componentOrTemplateRef
     * @return {?}
     */
    function (config, componentOrTemplateRef) {
        if (componentOrTemplateRef === void 0) { componentOrTemplateRef = MtxDialogComponent; }
        /** @type {?} */
        var data = Object.assign({}, defaults, config);
        return this.dialog.open(componentOrTemplateRef, __assign(__assign({}, data), { data: data }));
    };
    /**
     * @param {?} title
     * @param {?=} onOk
     * @return {?}
     */
    MtxDialog.prototype.alert = /**
     * @param {?} title
     * @param {?=} onOk
     * @return {?}
     */
    function (title, onOk) {
        if (onOk === void 0) { onOk = (/**
         * @return {?}
         */
        function () { }); }
        this.open({
            title: title,
            buttons: [
                {
                    type: 'warn',
                    text: 'OK',
                    onClick: (/**
                     * @return {?}
                     */
                    function () { return onOk(); }),
                },
            ],
            disableClose: true,
        });
    };
    /**
     * @param {?} title
     * @param {?=} onOk
     * @param {?=} onClose
     * @return {?}
     */
    MtxDialog.prototype.confirm = /**
     * @param {?} title
     * @param {?=} onOk
     * @param {?=} onClose
     * @return {?}
     */
    function (title, onOk, onClose) {
        if (onOk === void 0) { onOk = (/**
         * @return {?}
         */
        function () { }); }
        if (onClose === void 0) { onClose = (/**
         * @return {?}
         */
        function () { }); }
        this.open({
            title: title,
            buttons: [
                {
                    type: '',
                    text: 'CLOSE',
                    onClick: (/**
                     * @return {?}
                     */
                    function () { return onClose(); }),
                },
                {
                    type: 'warn',
                    text: 'OK',
                    onClick: (/**
                     * @return {?}
                     */
                    function () { return onOk(); }),
                },
            ],
            disableClose: true,
        });
    };
    MtxDialog.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    MtxDialog.ctorParameters = function () { return [
        { type: MatDialog }
    ]; };
    return MtxDialog;
}());
export { MtxDialog };
if (false) {
    /** @type {?} */
    MtxDialog.prototype.dialog;
}
export { ɵ0, ɵ1 };
