// ngx-color/github/esm2015/github-swatch.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, } from '@angular/core';
let GithubSwatchComponent = class GithubSwatchComponent {
    constructor() {
        this.onClick = new EventEmitter();
        this.onSwatchHover = new EventEmitter();
    }
    handleClick({ hex, $event }) {
        this.onClick.emit({ hex, $event });
    }
};
__decorate([
    Input()
], GithubSwatchComponent.prototype, "color", void 0);
__decorate([
    Output()
], GithubSwatchComponent.prototype, "onClick", void 0);
__decorate([
    Output()
], GithubSwatchComponent.prototype, "onSwatchHover", void 0);
GithubSwatchComponent = __decorate([
    Component({
        selector: 'color-github-swatch',
        template: `
  <div class="github-swatch">
    <color-swatch
      [color]="color"
      [style]="swatchStyle"
      (onClick)="handleClick($event)"
      (onHover)="onSwatchHover.emit($event)"
      class="swatch"
    ></color-swatch>
    <div class="clear"></div>
  </div>
  `,
        changeDetection: ChangeDetectionStrategy.OnPush,
        preserveWhitespaces: false,
        styles: [`
    .github-swatch {
      width: 25px;
      height: 25px;
      font-size: 0;
    }
    .github-swatch:hover {
      position: relative;
      z-index: 2;
      outline: 2px solid #fff;
      box-shadow: 0 0 5px 2px rgba(0,0,0,0.25);
    }
  `]
    })
], GithubSwatchComponent);
export { GithubSwatchComponent };
