// ngx-color/slider/esm5/slider.component.js
import { __decorate, __extends } from "tslib";
import { CommonModule } from '@angular/common';
import { ChangeDetectionStrategy, Component, Input, NgModule, } from '@angular/core';
import { ColorWrap, HueModule, SwatchModule } from 'ngx-color';
import { SliderSwatchComponent } from './slider-swatch.component';
import { SliderSwatchesComponent } from './slider-swatches.component';
var SliderComponent = /** @class */ (function (_super) {
    __extends(SliderComponent, _super);
    function SliderComponent() {
        var _this = _super.call(this) || this;
        _this.pointer = {
            width: '14px',
            height: '14px',
            borderRadius: '6px',
            transform: 'translate(-7px, -2px)',
            backgroundColor: 'rgb(248, 248, 248)',
            boxShadow: '0 1px 4px 0 rgba(0, 0, 0, 0.37)',
        };
        _this.radius = 2;
        return _this;
    }
    SliderComponent.prototype.handlePickerChange = function (_a) {
        var data = _a.data, $event = _a.$event;
        this.handleChange(data, $event);
    };
    __decorate([
        Input()
    ], SliderComponent.prototype, "pointer", void 0);
    __decorate([
        Input()
    ], SliderComponent.prototype, "radius", void 0);
    SliderComponent = __decorate([
        Component({
            selector: 'color-slider',
            template: "\n  <div class=\"slider-picker {{ className }}\">\n    <div class=\"slider-hue\">\n      <color-hue\n        [hsl]=\"hsl\" [radius]=\"radius\" [pointer]=\"pointer\"\n        (onChange)=\"handlePickerChange($event)\"\n      ></color-hue>\n    </div>\n    <div class=\"slider-swatches\">\n      <color-slider-swatches [hsl]=\"hsl\"\n        (onClick)=\"handlePickerChange($event)\"\n      ></color-slider-swatches>\n    </div>\n  </div>\n  ",
            changeDetection: ChangeDetectionStrategy.OnPush,
            preserveWhitespaces: false,
            styles: ["\n    .slider-hue {\n      height: 12px;\n      position: relative;\n    }\n  "]
        })
    ], SliderComponent);
    return SliderComponent;
}(ColorWrap));
export { SliderComponent };
var ColorSliderModule = /** @class */ (function () {
    function ColorSliderModule() {
    }
    ColorSliderModule = __decorate([
        NgModule({
            declarations: [
                SliderComponent,
                SliderSwatchComponent,
                SliderSwatchesComponent,
            ],
            exports: [SliderComponent, SliderSwatchComponent, SliderSwatchesComponent],
            imports: [CommonModule, HueModule, SwatchModule],
        })
    ], ColorSliderModule);
    return ColorSliderModule;
}());
export { ColorSliderModule };
