// ngx-color/block/esm5/block-swatches.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
var BlockSwatchesComponent = /** @class */ (function () {
    function BlockSwatchesComponent() {
        this.onClick = new EventEmitter();
        this.onSwatchHover = new EventEmitter();
    }
    BlockSwatchesComponent.prototype.ngOnInit = function () {
        this.swatchStyle = {
            width: '22px',
            height: '22px',
            float: 'left',
            marginRight: '10px',
            marginBottom: '10px',
            borderRadius: '4px',
        };
    };
    BlockSwatchesComponent.prototype.handleClick = function (_a) {
        var hex = _a.hex, $event = _a.$event;
        this.onClick.emit({ hex: hex, $event: $event });
    };
    BlockSwatchesComponent.prototype.focusStyle = function (c) {
        return {
            boxShadow: c + " 0 0 4px",
        };
    };
    __decorate([
        Input()
    ], BlockSwatchesComponent.prototype, "colors", void 0);
    __decorate([
        Output()
    ], BlockSwatchesComponent.prototype, "onClick", void 0);
    __decorate([
        Output()
    ], BlockSwatchesComponent.prototype, "onSwatchHover", void 0);
    BlockSwatchesComponent = __decorate([
        Component({
            selector: 'color-block-swatches',
            template: "\n    <div class=\"block-swatches\">\n      <color-swatch\n        *ngFor=\"let c of colors\"\n        [color]=\"c\"\n        [style]=\"swatchStyle\"\n        [focusStyle]=\"focusStyle(c)\"\n        (onClick)=\"handleClick($event)\"\n        (onHover)=\"onSwatchHover.emit($event)\"\n      ></color-swatch>\n      <div class=\"clear\"></div>\n    </div>\n  ",
            styles: ["\n    .block-swatches {\n      margin-right: -10px;\n    }\n    .clear {\n      clear: both;\n    }\n  "]
        })
    ], BlockSwatchesComponent);
    return BlockSwatchesComponent;
}());
export { BlockSwatchesComponent };
