// ngx-color/shade/fesm5/ngx-color-shade.js
import { __extends, __decorate } from 'tslib';
import { CommonModule } from '@angular/common';
import { Input, Component, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { toState, ColorWrap, ShadeModule } from 'ngx-color';

var ShadeSliderComponent = /** @class */ (function (_super) {
    __extends(ShadeSliderComponent, _super);
    function ShadeSliderComponent() {
        var _this = _super.call(this) || this;
        /** Pixel value for picker width */
        _this.width = 316;
        /** Pixel value for picker height */
        _this.height = 16;
        _this.pointer = {
            width: '18px',
            height: '18px',
            borderRadius: '50%',
            transform: 'translate(-9px, -2px)',
            boxShadow: '0 1px 4px 0 rgba(0, 0, 0, 0.37)',
        };
        return _this;
    }
    ShadeSliderComponent.prototype.ngOnChanges = function () {
        this.setState(toState(this.color, this.oldHue));
    };
    ShadeSliderComponent.prototype.handlePickerChange = function (_a) {
        var data = _a.data, $event = _a.$event;
        this.handleChange(data, $event);
    };
    __decorate([
        Input()
    ], ShadeSliderComponent.prototype, "width", void 0);
    __decorate([
        Input()
    ], ShadeSliderComponent.prototype, "height", void 0);
    ShadeSliderComponent = __decorate([
        Component({
            selector: 'color-shade-picker',
            template: "\n    <div class=\"shade-slider {{ className || '' }}\"\n      [style.width.px]=\"width\" [style.height.px]=\"height\">\n      <color-shade\n        [hsl]=\"hsl\"\n        [rgb]=\"rgb\"\n        [pointer]=\"pointer\"\n        (onChange)=\"handlePickerChange($event)\"\n      ></color-shade>\n    </div>\n  ",
            changeDetection: ChangeDetectionStrategy.OnPush,
            preserveWhitespaces: false,
            styles: ["\n    .shade-slider {\n      position: relative;\n    }\n  "]
        })
    ], ShadeSliderComponent);
    return ShadeSliderComponent;
}(ColorWrap));
var ColorShadeModule = /** @class */ (function () {
    function ColorShadeModule() {
    }
    ColorShadeModule = __decorate([
        NgModule({
            declarations: [ShadeSliderComponent],
            exports: [ShadeSliderComponent],
            imports: [CommonModule, ShadeModule],
        })
    ], ColorShadeModule);
    return ColorShadeModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { ColorShadeModule, ShadeSliderComponent };
