// ngx-color/esm2015/swatch.component.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { ChangeDetectionStrategy, Component, EventEmitter, Input, NgModule, Output, } from '@angular/core';
import { CheckboardModule } from './checkboard.component';
let SwatchComponent = class SwatchComponent {
    constructor() {
        this.style = {};
        this.focusStyle = {};
        this.onClick = new EventEmitter();
        this.onHover = new EventEmitter();
        this.divStyles = {};
        this.focusStyles = {};
        this.inFocus = false;
    }
    ngOnInit() {
        this.divStyles = Object.assign({ background: this.color, height: '100%', width: '100%', cursor: 'pointer', position: 'relative', outline: 'none' }, this.style);
    }
    currentStyles() {
        this.focusStyles = Object.assign(Object.assign({}, this.divStyles), this.focusStyle);
        return this.focus || this.inFocus ? this.focusStyles : this.divStyles;
    }
    handleFocusOut() {
        this.inFocus = false;
    }
    handleFocus() {
        this.inFocus = true;
    }
    handleHover(hex, $event) {
        this.onHover.emit({ hex, $event });
    }
    handleClick(hex, $event) {
        this.onClick.emit({ hex, $event });
    }
};
__decorate([
    Input()
], SwatchComponent.prototype, "color", void 0);
__decorate([
    Input()
], SwatchComponent.prototype, "style", void 0);
__decorate([
    Input()
], SwatchComponent.prototype, "focusStyle", void 0);
__decorate([
    Input()
], SwatchComponent.prototype, "focus", void 0);
__decorate([
    Output()
], SwatchComponent.prototype, "onClick", void 0);
__decorate([
    Output()
], SwatchComponent.prototype, "onHover", void 0);
SwatchComponent = __decorate([
    Component({
        selector: 'color-swatch',
        template: `
    <div
      class="swatch"
      [ngStyle]="currentStyles()"
      [attr.title]="color"
      (click)="handleClick(color, $event)"
      (keydown.enter)="handleClick(color, $event)"
      (focus)="handleFocus()"
      (blur)="handleFocusOut()"
      (mouseover)="handleHover(color, $event)"
      tabindex="0"
    >
      <ng-content></ng-content>
      <color-checkboard
        *ngIf="color === 'transparent'"
        boxShadow="inset 0 0 0 1px rgba(0,0,0,0.1)"
      ></color-checkboard>
    </div>
  `,
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], SwatchComponent);
export { SwatchComponent };
let SwatchModule = class SwatchModule {
};
SwatchModule = __decorate([
    NgModule({
        declarations: [SwatchComponent],
        exports: [SwatchComponent],
        imports: [CommonModule, CheckboardModule],
    })
], SwatchModule);
export { SwatchModule };
